File: zerog_scotlandyard/GameLogic.java
package zerog_scotlandyard;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class GameLogic {
	final static int MAX_MRX_MOVES = 24;
	static Map<PlayerID, Player> players = new HashMap<PlayerID, Player>();
	static Map <NodeID, Node> board = new HashMap<NodeID, Node>();

	//read in line into a string
	public void BuildMap(string textFile){
		String line = br.readLine;
		while (line != null){
			line.split(',');	//split string, third element is transportType
			addLines(line, line[2])
		}
	}
	
	//add connections to map
	public void addLines (string line, string type){
		//use inter.parseInt() to get the integer values
		switch(type){
			case taxi:
				board.get(Integer.parseInt(line[0]).line(board.get(Integer.parseInt(line[1])), taxi);
				board.get(Integer.parseInt(line[1]).line(board.get(Integer.parseInt(line[0])), taxi);
			case bus:
				board.get(Integer.parseInt(line[0]).line(board.get(Integer.parseInt(line[1])), bus);
				board.get(Integer.parseInt(line[1]).line(board.get(Integer.parseInt(line[0])), bus);
			case underground:
				board.get(Integer.parseInt(line[0]).line(board.get(Integer.parseInt(line[1])), underground);
				board.get(Integer.parseInt(line[1]).line(board.get(Integer.parseInt(line[0])), underground);
		}
	}
	
	/*
	for (i = 0; i < numLines, i++){		//still working on it
		addLines (taxiLines[i][], Taxi);
		}
	addLines (busLines[i][], Bus);
	addLines (ugLines[i][], Underground);
	*/
	
	
	
	
	public static boolean canMove (Player player) {
		Set<Node> allPossibleMoveLocations = player.getLocation().getAllEdges();
		Set<Node> allOtherPlayerLocations = getAllOtherPlayerLocations(player);
		allPossibleMoveLocations.removeAll(allOtherPlayerLocations);
		boolean playerBlocked = allPossibleMoveLocations.isEmpty();

		if(playerBlocked){
			return false;
		}

		for(Node adjacentLocation : allPossibleMoveLocations){
			Set<TransportType> connectionTypes = getPossibleConnectionTypes(player.getLocation(), adjacentLocation);
			for(TransportType transportType : connectionTypes){
				if(playerHasEnoughTickets(player, transportType)){
					return true;
				}
			}
		}
		//Player does not have enough tickets
		return false;
	}


	/**
	 * If the set is empty, there is no connection. If the set contains at least one transportType, those
	 * are the tickets you can use to get to the destination.
	 */
	public static Set<TransportType> getPossibleConnectionTypes(Node current, Node destination){
		Set<TransportType> usableTickets = new HashSet<TransportType>();
		for(TransportType type : TransportType.values()){
			Set<Node> edges = current.getEdges(type);
			if(edges.contains(destination)){
				usableTickets.add(type);
			}
		}
		return usableTickets;
	}

	public static boolean playerHasEnoughTickets(Player player, TransportType transportType){
		return (player.tickets.get(transportType).intValue() > 0);
	}

	public static Set<Node> getAllOtherPlayerLocations(Player currentPlayer){
		Set<Node> locations = new HashSet<Node>();
		for(Player player : players.values()){
			if (player != currentPlayer){
				locations.add(player.getLocation());
			}
		}
		return locations;
	}
	
	
	public boolean checkWin() {
		for (Player player : players.value()) {
			if (player instanceof Detective) {
				if (player.getLocation() == players.get(0).getLocation())	//assuming Mrx "key" or PlayerID is 0
					return true;
			}
			else if (player instanceof MrX) {
				int CantMoves = 0;
				for (Player detective : players.values()) {
					if (!canMove(detective))
						CantMoves++;
				}
				if (CantMoves == (players.size()-1))
					return true;
				if (player.getMoveLogSize() == MAX_MRX_MOVES)
					return true;
			}
			else 
				throw new IllegalArgumentException();	//throw exception
		}
		return false;
	}
}

File: tests/Tests/GameLogic.cs
using System.Collections;
using System.Collections.Generic;
using System;
using System.IO;
using System.Linq;

public class GameLogic {

	static readonly int MAX_MRX_MOVES = 24;
	public static GamePosition GamePosition { get; set; }

	public static void BuildMap(string textFile){
				//create array of strings to hold lines of text file
				string [] lines = File.ReadAllLines(textFile);

				//loop through each line (string) 
				foreach (string line in lines){
					//split it into an an array of strings 
					string [] connection = line.Split(',');

					//create new nodes
					int nodeID0 = int.Parse (connection [0]);
					int nodeID1 = int.Parse (connection [1]);
					TransportType connType;

					switch (connection [2]) {
						case "taxi":
							connType = TransportType.taxi;
							break;
						case "bus":
							connType = TransportType.bus;
							break;
						case "underground":
							connType = TransportType.underground;
							break;
						default:
							throw new Exception();
					}

					//check if nodes are already in board
					if (!GamePosition.Board.ContainsKey(nodeID0)) {
						GamePosition.Board.Add(nodeID0, new Node(nodeID0));
					}
					if (!GamePosition.Board.ContainsKey(nodeID1)) {
						GamePosition.Board.Add(nodeID1, new Node(nodeID1));
					}

					GamePosition.Board[nodeID0].addEdges(GamePosition.Board[nodeID1], connType);
				}
		}

	public static bool canMove (Player player) {
		HashSet<Node> allPossibleMoveLocations = new HashSet<Node>();

		allPossibleMoveLocations = player.getLocation().getAllEdges();
		allPossibleMoveLocations.ExceptWith (getAllOtherPlayerLocations (player));

		bool playerBlocked = (allPossibleMoveLocations.Count == 0);
		
		if(playerBlocked){
			return false;
		}

		foreach (Node adjacentLocation in allPossibleMoveLocations) {
			//create new HashSet that will hold the connection types... not sure if inside or outside loop.
			HashSet<TransportType> connectionTypes = getPossibleConnectionTypes(player.getLocation(), adjacentLocation);
			foreach(TransportType type in connectionTypes){
				if(playerHasEnoughTickets(player, type)){
					return true;
				}
			}
		}

		//player does not have enough tickets
		return false;
	}

	public static HashSet<Node> getAllOtherPlayerLocations(Player currentPlayer){
		HashSet<Node> locations = new HashSet<Node>();
		foreach (Player player in GamePosition.Players.Values){
			if (player != currentPlayer){
				locations.Add(player.getLocation());
			}
		}
		return locations;
	}


	public static HashSet<TransportType> getPossibleConnectionTypes (Node current, Node destination){
		HashSet<TransportType> usableTickets = new HashSet<TransportType> ();
		foreach(TransportType type in Enum.GetValues(typeof(TransportType))){
			HashSet<Node> edges = current.getEdges(type);
			if(edges.Contains(destination))
				usableTickets.Add(type);
		}
		return usableTickets;
	}

	public static bool playerHasEnoughTickets(Player player, TransportType transportType){
		return (player.getTickets(transportType) > 0);
	}

	public static bool checkWin(Dictionary<int, Player> players, out int winningPlayerId) {
		foreach (Player player in players.Values){
			if (player is Detective){
				if (player.getLocation() == players[0].getLocation()){
					winningPlayerId = player.getId ();
					return true;
				}
			} else if (player is MrX){
				int CantMoves = 0;
				foreach(Player detective in players.Values) {
					if (!canMove(detective))
						CantMoves++;
				}
				if (CantMoves == (players.Count-1)){
					winningPlayerId = player.getId ();
					return true;
				}
				if (player.getMoveLogSize() == MAX_MRX_MOVES){
					winningPlayerId = player.getId ();
					return true;
				}
				
			} else {
					throw new ArgumentException();
			}
		}
		winningPlayerId = -1;
		return false;
	}

	public static GamePosition ApplyMove(int player, Node move, GamePosition gamePosition){

		GamePosition potentialResult = gamePosition;

		HashSet<TransportType> possibleConnection = getPossibleConnectionTypes(potentialResult.Players[player].getLocation(), move);

		foreach (TransportType type in possibleConnection) {
			if (playerHasEnoughTickets(potentialResult.Players[player], type)){
				potentialResult.Players[player].move(move, type);
				break;
			}
		}
		return potentialResult;
	}

	public static HashSet<Node> GetPossibleMoves(int player){
		Node currentLocation = GamePosition.Players[player].getLocation();
		HashSet<Node> possibleLocations = new HashSet<Node>();
		HashSet<Node> occupiedLocations = GameLogic.getAllOtherPlayerLocations(GamePosition.Players[player]);
		HashSet<Node> checkLocations;

		if(playerHasEnoughTickets(GamePosition.Players[player], TransportType.blackCard)){
			checkLocations = currentLocation.getAllEdges();
			checkLocations.ExceptWith(occupiedLocations);
			possibleLocations.Union(checkLocations);
		} else {
			foreach(TransportType type in Enum.GetValues(typeof(TransportType))){
				if(playerHasEnoughTickets(GamePosition.Players[player], type)){
					checkLocations = currentLocation.getEdges(type);
					checkLocations.ExceptWith(occupiedLocations);
					possibleLocations.Union(checkLocations);
				}
			}
		}
		return possibleLocations;
	}
}